File: 010.regular-expression-matching/regular-expression-matching2.js
/**
 * @param {string} s
 * @param {string} p
 * @return {boolean}
 */
// memo 自底而上
var isMatch = function (s, p) {
  let memo = { [`${s.length},${p.length}`]: true }
  for (let i = 0; i < s.length; i++) {
    memo[`${i},${p.length}`] = false
  }
  for (let i = s.length; i > -1; i--) {
    for (let j = p.length - 1; j > -1; j--) {
      firstMatch = i < s.length && (p[j] === '.' || p[j] === s[i])
      if (j + 1 < p.length && p[j + 1] === '*') {
        memo[`${i},${j}`] = memo[`${i},${j + 2}`] || firstMatch && memo[`${i + 1},${j}`]
      }
      else if (firstMatch) {
        memo[`${i},${j}`] = memo[`${i + 1},${j + 1}`]
      }
      else {
        memo[`${i},${j}`] = false
      }
    }
  }
  return memo['0,0']
};

console.log(isMatch("aa", "a") == false)
console.log(isMatch("aa", "a*") == true)
console.log(isMatch("ab", ".*") == true)
console.log(isMatch("fghijkae", ".*ae") == true)
console.log(isMatch("aab", "c*a*b") == true)
console.log(isMatch("mississippi", "mis*is*p*.") == false)
console.log(isMatch("aaa", "a*a") == true)
console.log(isMatch("aaa", "ab*a*c*a") == true)
console.log(isMatch("abbabaaaaaaacaa", "a*.*b.a.*c*b*a*c*") == true)
console.log(isMatch("bcbabcaacacbcabac", "a*c*a*b*.*aa*c*a*a*") == true)

File: hard/010.regular-expression-matching/regular-expression-matching2.py
# 解法2：根据第二个位置情况分类
class Solution:

    def isMatch(self, s: str, p: str) -> bool:
        if not p:
            return not bool(s)
        if not s:
            return len(p) % 2 == 0 and all([i == '*' for i in p[1::2]])

        firstMatch = p[0] == s[0] or p[0] == '.'

        if len(p) > 1 and p[1] == '*':
            return self.isMatch(s, p[2:]) or firstMatch and self.isMatch(s[1:], p)
        if firstMatch:
            return self.isMatch(s[1:], p[1:])
        return False


if __name__ == '__main__':
    s = Solution()
    isMatch = s.isMatch

    print(isMatch("aa", "a") == False)
    print(isMatch("aa", "a*") == True)
    print(isMatch("ab", ".*") == True)
    print(isMatch("fghijkae", ".*ae") == True)
    print(isMatch("aab", "c*a*b") == True)
    print(isMatch("mississippi", "mis*is*p*.") == False)
    print(isMatch("aaa", "a*a") == True)
    print(isMatch("aaa", "ab*a*c*a") == True)
    print(isMatch("abbabaaaaaaacaa", "a*.*b.a.*c*b*a*c*") == True)
    print(isMatch("bcbabcaacacbcabac", "a*c*a*b*.*aa*c*a*a*") == True)
